File: library/src/main/java/com/proxerme/library/api/HttpUrlAdapter.java
package com.proxerme.library.api;

import com.squareup.moshi.FromJson;
import com.squareup.moshi.ToJson;
import okhttp3.HttpUrl;

/**
 * @author Ruben Gees
 */
final class HttpUrlAdapter {

    @FromJson
    HttpUrl fromJson(final String url) {
        final String fixedUrl = url.startsWith("//") ? "http:" + url : url;

        return HttpUrl.parse(fixedUrl);
    }

    @ToJson
    String toJson(final HttpUrl url) {
        return url.toString();
    }
}

File: library/src/main/kotlin/me/proxer/library/internal/adapter/HttpUrlAdapter.kt
package me.proxer.library.internal.adapter

import com.squareup.moshi.FromJson
import com.squareup.moshi.ToJson
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull

/**
 * @author Ruben Gees
 */
internal class HttpUrlAdapter {

    @FromJson
    fun fromJson(url: String): HttpUrl? {
        val fixedUrl = if (url.startsWith("//")) "http:$url" else url

        return fixedUrl.toHttpUrlOrNull()
    }

    @ToJson
    fun toJson(url: HttpUrl?) = url?.toString()
}
